### PR TITLE
GEFF IO improvement

### PR DESCRIPTION
**Pull request by JoOkuma**
Hi @cmalinmayor 

# Proposed Change

Optimizes zarr I/O in `core_io` by replacing boolean-mask indexing with `oindex` integer indexing for reads, and adding explicit chunking with zarr v3 sharding for writes.

While profiling `tracksdata`, I noticed two performance issues:

1. `_load_prop_to_memory` converts boolean masks to Python lists via `mask.tolist()` before indexing zarr arrays. For sparse selections (e.g. 5 nodes out of 8674) this is ~7x slower than using integer indices with `oindex`, because zarr has to process the entire boolean array instead of just reading the relevant chunks.

2. No explicit chunking when writing arrays — zarr's auto-chunking splits trailing dimensions unnecessarily. For example, `edge_ids` with shape `(189870, 2)` gets chunked as `(47468, 1)`, putting each column in a separate chunk.

Also fixes a bug: the per-node boolean mask was applied to the varlength `DATA` array, which is a flat 1D concatenation with a completely different length. This crashes on zarr v3 with `VindexInvalidSelectionError`. The DATA array must always be loaded in full since deserialization uses byte offsets from VALUES.

### Benchmarks

Dataset: BF-C2DL-HSC/01 (8674 nodes, 189870 edges)

**`reader.build()` selecting 5 nodes:**

| Version | Time |
|---------|------|
| Before (`mask.tolist`) | ~155 ms |
| After (`oindex`) | ~21 ms |

**File count on zarr v3 write:**

| Version | Files on disk |
|---------|--------------|
| Before (auto-chunking) | 119 |
| After (sharding) | 67 |

**Chunking examples:**

| Array | Before | After |
|-------|--------|-------|
| `edge_ids` (189870, 2) uint64 | (47468, **1**) | (189870, **2**) |
| `bbox` (8674, 6) int64 | (4337, 6) | (8674, 6) |
| `mask/data` (4192807,) uint64 | (131026,) | (1048576,) |

# Types of Changes

- Bugfix (non-breaking change which fixes an issue)
- New feature or enhancement

Which topics does your change affect?

- Core io

# Checklist

- [x] I have read the [developer/contributing](https://github.com/live-image-tracking-tools/geff/blob/main/CONTRIBUTING) docs.
- [ ] I have added tests that prove that my feature works in various situations or tests the bugfix (if appropriate).
- [x] I have checked that I maintained or improved code coverage.
- [x] I have written docstrings and checked that they render correctly by looking at the docs preview (link left as a comment on the PR).

# Further Comments

The change is split into two commits:

1. **Reading**: adds `_load_zarr_subset()` using `oindex` and fixes the varlength DATA bug. The boolean mask API of `_load_prop_to_memory` and `build()` is unchanged — the conversion to integer indices happens internally.

2. **Writing**: adds `_write_zarr_array()` that chunks only along the first dimension (power-of-two size, ~8 MiB target). On zarr v3 it uses sharding so each array becomes a single file with sub-chunks inside. On zarr v2 it just uses the larger explicit chunks. Empty arrays fall back to direct assignment.

All 116 `test_core_io` tests pass.
I haven't added new tests yet — the existing masked-read tests (`test_build_w_masked_nodes`, `test_load_prop_into_memory`, etc.) already exercise the new code paths since the public API is unchanged.
Happy to add targeted benchmarks or edge-case tests if needed.

### CODE REVIEW

File: packages/geff/src/geff/core_io/_base_write.py [hunk 1]
```diff
@@ -307,6 +307,74 @@ def write_arrays(
...
+    import zarr as _zarr
+
+    if data.shape[0] == 0:
+        group[path] = data
+        return
+
+    first_dim = _compute_first_dim_chunk(data.shape, data.dtype.itemsize)
+    chunks = (first_dim, *data.shape[1:])
+
+    use_sharding = (
```
Comment (cmalinmayor) [inline packages/geff/src/geff/core_io/_base_write.py]: We have a zarr utils function for detecting the zarr format version in core_io/utils.py - could use here, and it also works on any store not just the group

File: packages/geff/src/geff/core_io/_base_write.py [hunk 2]
```diff
@@ -307,6 +307,74 @@ def write_arrays(
...
+    row_bytes = int(np.prod(shape[1:])) * itemsize if len(shape) > 1 else itemsize
+    n_rows = _TARGET_CHUNK_BYTES // max(row_bytes, 1)
+    # Round down to nearest power of two
+    if n_rows >= 1:
+        n_rows = 1 << (n_rows.bit_length() - 1)
+    return max(1, min(first_dim, n_rows))
+
+
+def _write_zarr_array(
+    group: Any,
```
Comment (cmalinmayor) [inline packages/geff/src/geff/core_io/_base_write.py]: What is thea ctual required type? Is it a general StoreLike, or does it need to be zarr.group?

**Review by cmalinmayor — CHANGES_REQUESTED** [review]
Generally looks really good, thanks! I would love a test for the bug that would have failed on main and succeeded here. If I get the chance today I'll add that, otherwise if you beat me to it thats great as well.  Also, a minor comment about our zarr utilities/typing.

File: packages/geff/tests/test_core_io/test_base_read.py [hunk 3]
```diff
@@ -226,6 +226,38 @@ def test_read_edge_props() -> None:
     _ = NxBackend.construct(**in_memory_geff)
 
 
+def test_build_w_masked_nodes_varlength() -> None:
```
Comment (JoOkuma) [inline packages/geff/tests/test_core_io/test_base_read.py:255]: Thanks for the review @cmalinmayor . This is the regression test you mentioned above. Nonetheless, don't hesitate to add more tests.

I'll address your comments soon.

File: packages/geff/src/geff/core_io/_base_write.py [hunk 4]
```diff
@@ -302,6 +302,74 @@ def write_arrays(
...
+    On zarr v3 sharding is used so every array becomes a single file: the shard
+    (outer chunk) spans the entire first dimension while inner sub-chunks are
+    ~8 MiB each.
+
+    Args:
+        group: The zarr group to write into.
+        path: The path within the group.
+        data: The numpy array to write.
+    """
+    import zarr as _zarr
```
Comment (cmalinmayor) [inline packages/geff/src/geff/core_io/_base_write.py]: Why re-importing this here? Can we just use the import at the top?

**Comment by cmalinmayor** [comment]
Added a test just to make coverage happy since technically it was correct that we didn't test those two lines... 🙃 Otherwise you are correct the tests are great! Just missed it somehowd

File: packages/geff/src/geff/core_io/_base_write.py (hunk 4 above)
Comment (JoOkuma) [reply packages/geff/src/geff/core_io/_base_write.py]: yes

File: packages/geff/src/geff/core_io/_base_write.py (hunk 2 above)
Comment (JoOkuma) [reply packages/geff/src/geff/core_io/_base_write.py]: It's a zarr.Group, it doesn't work with `StoreLike` because it can be indexed by `path`.

**Comment by JoOkuma** [comment]
Hi @cmalinmayor, I addressed your comments, lmk if I missed anything

File: packages/geff/src/geff/core_io/_base_write.py [hunk 5]
```diff
@@ -302,6 +304,70 @@ def write_arrays(
...
+    row_bytes = int(np.prod(shape[1:])) * itemsize if len(shape) > 1 else itemsize
+    n_rows = _TARGET_CHUNK_BYTES // max(row_bytes, 1)
+    # Round down to nearest power of two
+    if n_rows >= 1:
+        n_rows = 1 << (n_rows.bit_length() - 1)
+    return max(1, min(first_dim, n_rows))
+
+
+def _write_zarr_array(
+    group: Any,
```
Comment (cmalinmayor) [inline packages/geff/src/geff/core_io/_base_write.py]:
```suggestion
    group: zarr.Group,
```

**Review by cmalinmayor — APPROVED** [review]
Thank you!! 🦖 🍅 🎉
